### PR TITLE
Feed Recorded Future download all compressed data on disk bug 

### PR DESCRIPTION
**Pull request by anas-yousef**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/XSUP-29744

## Description
Added the support to stream incoming gzip-compressed data, and save them in chunks, without the need to load all the data on disk at once.

## Must have
- [ ] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py [hunk 1]
```diff
@@ -159,26 +158,84 @@ def build_iterator(self, service, indicator_type, risk_rule: Optional[str] = Non
                              " requests made to Recorded Future. ")
             else:
                 return_error(
-                    '{} - exception in request: {} {}'
-                    .format(self.SOURCE_NAME, response.status_code, response.content))  # type: ignore
+                    f'{self.SOURCE_NAME} - exception in request: {response.status_code} {response.content}'
+                )  # type: ignore
```
[the remark this answers is not in the document]
Comment (anas-yousef) [reply Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py]: NVM, I just needed to move the `# type: ignore` to its original position.

File: Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py [hunk 2]
```diff
@@ -159,26 +158,84 @@ def build_iterator(self, service, indicator_type, risk_rule: Optional[str] = Non
...
+        The decoding of the bytes was outsourced to this function, in order to test the decoding mechanism
+        in the 'stream_compressed_data' function.
+        """
+        return bytes_to_decode.decode(encoding=encoding)
+
     def get_batches_from_file(self, limit):
         demisto.info('reading from file')
         # we do this try to make sure the file gets deleted at the end
         try:
-            file_stream = open("response.txt", 'rt')
```
Comment (JasBeilin) [inline Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:181]: Is this on purpose?

**Review by JasBeilin — COMMENTED** [review]
Looks good but I do have some comments. Please check it.

File: Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py (hunk 2 above)
Comment (anas-yousef) [reply Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:181]: This was changed by ruff.

Comment (JasBeilin) [reply Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:181]: So please revert.

File: Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py [hunk 3]
```diff
@@ -159,26 +158,84 @@ def build_iterator(self, service, indicator_type, risk_rule: Optional[str] = Non
...
+                    chunks_counter += 1
+                    should_cut_off_bytes = True
+                    bytes_to_cut = 0
+                    decompressed_chunk = decompressor.decompress(chunk)
+                    if cut_off_bytes:
+                        # To concatenate cut off bytes from previous chunk
+                        decompressed_chunk = cut_off_bytes + decompressed_chunk
+                        cut_off_bytes = b''
+                    chunk_to_decode = decompressed_chunk
+                    while should_cut_off_bytes:
```
[the remark this answers is not in the document]
Comment (anas-yousef) [reply Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:207]: True, I changed it to use a for loop, saving some lines.

File: Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py [hunk 4]
```diff
@@ -159,21 +158,76 @@ def build_iterator(self, service, indicator_type, risk_rule: Optional[str] = Non
...
+            for chunk in response.iter_content(chunk_size):
+                if chunk:
+                    chunks_counter += 1
+                    decompressed_chunk = decompressor.decompress(chunk)
+                    if cut_off_bytes:
+                        # To concatenate cut off bytes from previous chunk
+                        decompressed_chunk = cut_off_bytes + decompressed_chunk
+                        cut_off_bytes = b''
+                    chunk_to_decode = decompressed_chunk
+                    for bytes_to_cut in [1, 2, 3, 4]:
```
Comment (JasBeilin) [inline Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:205]:
```suggestion
                    for bytes_to_cut in [0, 1, 2, 3]:
```
first one should not remove bytes if not needed.

File: Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py [hunk 5]
```diff
@@ -159,21 +158,76 @@ def build_iterator(self, service, indicator_type, risk_rule: Optional[str] = Non
...
+                        try:
+                            decoded_counter += 1
+                            decoded_str = self.decode_bytes(chunk_to_decode)
+                            f.write(decoded_str)
+                            break
+                        except UnicodeDecodeError as decode_error:
+                            demisto.debug('Decoding using UTF-8 failed due to cut-off character while streaming.'
+                                          f' Will remove {bytes_to_cut} byte/s from the chunk and decode again.')
+                            if bytes_to_cut >= 4:
```
Comment (JasBeilin) [inline Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:214]: This should not happen in a for loop running to 3.

**Review by JasBeilin — COMMENTED** [review]
great.

File: Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py [hunk 6]
```diff
@@ -162,7 +162,7 @@ def build_iterator(self, service, indicator_type, risk_rule: str | None = None):
                 )
 
         if service == 'connectApi':
-            self.stream_compressed_data(response=response, chunk_size=CHUNK_SIZE)
+            self.stream_compressed_data(response=response, chunk_size=1024)
```
Comment (JasBeilin) [inline Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py:165]: Is this intended?

Comment (anas-yousef) [reply Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture/FeedRecordedFuture.py]: Nope, by accident, I fixed it

**Pull request merged**